### ISSUE TITLE
curious about issues with gcc compilers after 7.5.0?

### ISSUE THREAD

**Issue by bkerin** [igormiktor/build-avr-gcc#1]
I'm currently updating my avr toolchain and happened accross your project.  Have you heard of anyone using recent avr-gcc (e.g 10 or 12)?  I keep running into people saying "code quality declined later" etc. and I notice that official microchip etc. look like they're stuck on very old gcc.

What is your opinion of avr-gcc overall?  Is it effectively dead in new gcc or what?

**Comment by igormiktor** [comment]
So sorry for somehow missing this until today (from 4 years ago!).  Support for AVR in GCC from MicroChip ceased after 7.4.  There wasn't an official announcement but at some point in either reddit or in one of the MicroChip messaging boards, someone from MicroChip actually said that.  I'd have to google a bit to find it.  

GCC 7.5.0 as best I can tell still works just fine with the AVRMega series chips that I work with a lot.  And I suspect it works fine with the entire 8-bit AVR series.  However, every GCC version after 7.5.0 I've tested generates bad code (I've tested up through GCC 13.x).  

It often works on simple code, but with larger, complex code it fails to run (e.g., see my CARRTv2 repo).  I've never been able to reproduce the codegen bug in a simple, easy to understand example.  At some point, I just stopped trying.

So like MicroChip (who I believe is on 7.4.1?), I'm stuck on GCC 7.5.0 for all my AVR stuff.

Again, sincere apologies for missing your comment back when you originally posted it!

**Comment by igormiktor** [comment]
This post (https://forum.arduino.cc/t/please-provide-new-gcc-toolchaains/1230603) says GCC 7.3 " is the last version of gcc "supported" by the chip Vendor (Atmel, Microchip.) "

**Comment by bkerin** [comment]
Yikes.  I've been using much more recent versions though not super regularly for a while now and they *seem* to work.  I thought a while back there was a bit of an outcry about Microchip letting the AVR GCC die despite "nothing bad" happening to Atmel and it got a bounty and was supposedly fixed.  And it does seem to work for several non-toy projects for me.

This project: https://github.com/avrdudes/avr-libc

recommends to use the latest GCC

**Comment by igormiktor** [comment]
Interesting.  I may try this again. Thanks.

**Comment by igormiktor** [comment]
Following up, I've been experiment with the latest AVR-GCC versions -- thanks for pointing me to these (and the new avr-libc).

There is one thing that works in GCC 7.5 but not in any of the newer versions:  an interrupt routine to implement asynchronous I2C.  Code generated by GCC 8, 14, and 15.2 for this ISR all consistently lock up avr328 and avr2560 chips.  I'm assuming it's an error in my code, but...

1. can't find any coding error in careful line-by-line reviews or manually walking through executing of the I2C state engine
2.  when I debug the ISR using LED lights to confirm/deny correctness, the internal state of the ISR routine always looks 100% what is expected (right up until the avr locks up)   
3. the ISR works 100% with GCC versions 5 thru 7.5
4. Codegen looks distinctly different between 5-7.5 versions and GCC 8 and beyond (asm generated by versions 5-7.5 look nearly identical side-by-side; and same for versions 8 and beyond).  I haven't had time to go through the 7.5 and 15.2 generated asm to try to line things up and see what the actual differences are (versus simply that the asm text looks different on a quick side-by-side scan) and explore if any such differences might be the source of the problem.  I have noticed that in places order of asm statements differs, but not in ways that I can see impacting actual execution.

Apart from the one ISR (which unfortunately I rely on a lot), GCC 15 and latest avr-libc work well.

Whenever I figure this out, I'll update you

**Comment by bkerin** [comment]
I assume you're using an explicit volatile declaration for any variables used by both the ISR and other code?  I believe optimization probably changed quite a bit about GCC 8

**Comment by igormiktor** [comment]
Absolutely.  And ISR code that implements asynchronous UART transmission is passing all tests and working fine across all GCC versions.  And both the ISR for async I2C and UART have a common code base for the ring-buffer that stores incoming/outgoing msgs (this is where almost all of the interaction between ISR and regular code happens).  

Anyway, requires more focused thinking and debugging than I currently have time for

**Comment by bkerin** [comment]
If you have one of the examples that breaks things in sufficiently isolated form st it builds easily I'd like to take a look at it.  I still use 328p and 2560 for a lot of stuff and don't like the idea of broken ISRs

**Comment by igormiktor** [comment]
That's really kind of you.  Thanks for offering.

My current test configuration is a 328p (concretely an Arduino Uno) connected talking  to an LCD and keypad like [this one](https://www.adafruit.com/product/714).  The 328p talks i2c to the MCP23017 which then drives the LCD.    I use it because it is trivial to wire up and easy to verify correct operation. And it consistently locks up in the early stages of sending configuration commands via i2c to the device.

If you have access to an LCD device like that, next time I have an opportunity I'll generate a minimal self-contained, easy to build test case to exercise the i2c ISR that demonstrates the lock-up under GCC 15.2 but works  under GCC 7.5.

If you don't have access to such an LCD, let me know what i2c slave devices you have handy and if I have such a device I can build a test case using that.
